Storage/SQL/SQLiteHistory.swift

```diff
@@ -554,7 +554,27 @@ extension SQLiteHistory: BrowserHistory {
     }
 
     public func getSitesByLastVisit(_ limit: Int) -> Deferred<Maybe<Cursor<Site>>> {
-        return self.getFilteredSitesByVisitDateWithLimit(limit, whereURLContains: nil, includeIcon: true)
+        let sql = [
+            "SELECT",
+                "historyID, url, title, guid, domain_id, domain,",
+                "CASE is_local WHEN 1 THEN date ELSE 0 END AS localVisitDate,",
+                "CASE is_local WHEN 0 THEN date ELSE 0 END AS remoteVisitDate,",
+                "iconID, iconURL, iconDate, iconType, iconWidth",
+            "FROM (",
+                "SELECT",
+                    "\(TableHistory).id AS historyID, \(TableHistory).url, title, guid, domain_id, is_local, iconID, iconURL, iconDate, iconType, iconWidth, MAX(date) AS date",
+                "FROM \(TableHistory)",
+                "INNER JOIN \(TableVisits) ON \(TableVisits).siteID = \(TableHistory).id",
+                "LEFT OUTER JOIN \(ViewHistoryIDsWithWidestFavicons) ON \(ViewHistoryIDsWithWidestFavicons).id = \(TableHistory).id",
```

Comment:
I haven't run `EXPLAIN`, but we probably shouldn't be doing this inside the `GROUP BY`. If you do it inside you'll join icons against every visit; if you do it outside the query with `max(date)` then you'll reduce the number of rows you'll join. This is exactly what we do when joining against domains. Indeed, do it after we've `INNER JOIN`ed on domains!

And yes, flattening our icon stuff, as discussed on Slack, is probably a good idea.